---
title: "Mapping Example"
author: "Jeekin Lau"
date: "Updated: 11/29/2022"
output:   
  html_document:
    toc: true
    toc_float: 
      collapsed: false
---

# Introduction

This is a tutorial using a very clean dataset from array generated data on a well characterized dataset that has been published already [Rose Rosette Disease Resistance Loci Detected in Two Interconnected Tetraploid Garden Rose Populations](https://www.frontiersin.org/articles/10.3389/fpls.2022.916231/full)

To download the dataset please go to <https://github.com/jeekinlau/mapping_example/tree/main/data>

# Import the data

Download the dataset and import the dataset into R

```{r}
library(mappoly)
dat = read_geno_csv(file="data/BExMG_subset.csv", ploidy = 4)
```

# Filtering Data

## Filter by missing data

Typically, there will be many markers filtered out when filtered by marker, or by individual. Because this is a very well curated dataset, There are no markers filtered out. Start with 0.05 but there are no "rules" for what level to use.

```{r}
## Filtering dataset by marker
dat = filter_missing(input.data = dat, type = "marker", 
                      filter.thres = 0.05, inter = F)
## Filtering dataset by individual
dat = filter_missing(input.data = dat, type = "individual", 
                      filter.thres = 0.05, inter = F)
dat
```

## Filter by segregation distortion

Also very few markers removed due to good dataset. The figure below shows the marker types, the dosage of all the marker types, and the segregation distortion of the markers. Here also we apply an alpha=0.05 with Bonfferoni correction.

```{r}
## Filter according to segregation test
pval.bonf = 0.05/dat$n.mrk
mrks.chi.filt = filter_segregation(dat, chisq.pval.thres =  pval.bonf, inter = F)
seq.init = make_seq_mappoly(mrks.chi.filt)
seq.init

plot(seq.init)
```

# Calculating recombination fractions

This step allows for the pairwise recombination fraction calculation between all markers. If you computer has multiple cores use the `parallel::detectCores()` function to find out how many cores your local computer has. This step takes a while especially with a laptop with limited cores and RAM memory. (**there are tricks around this email me for alternative strategies around this**)

```{r}
## Accounting for recombination fractions

all.rf.pairwise = est_pairwise_rf(input.seq = seq.init, ncpus = 6)

```

## Plotting RF matrix

Plotting the RF matrix shows a heat map of markers that are closely linked versus those that are far away. Naturally we see that those markers group into the 7 chromosomes of the Rose genome.

```{r}
mat = rf_list_to_matrix(all.rf.pairwise)
plot(mat)
```

# Grouping

Next step is to group the markers into a list containing the necessary objects needed for mapping (ie markers and their two point rf estimate and a sequence of markers). There are several ways to approach this. Below we compare the grouping using UPGMA method, genomic (physical position according to genome position), and an intersect between the two.

```{r}
## Getting group information (UPGMA)
grs = group_mappoly(input.mat = mat,
                     expected.groups = 7,
                     comp.mat = TRUE, 
                     inter = F)
grs
plot(grs)
```

```{r}
## Making groups with intersection between UPGMA (rf) + genomic information (discards scaffolds)

LGS.inter=vector("list", 7)
for(j in 1:7){
  temp1 = make_seq_mappoly(grs, j, genomic.info=1)
  tpt = make_pairs_mappoly(all.rf.pairwise, input.seq = temp1)
  temp2 = rf_snp_filter(input.twopt = tpt, diagnostic.plot = FALSE)
  tpt2 = make_pairs_mappoly(tpt, input.seq = temp2)
  LGS.inter[[as.numeric(names(table(temp2$chrom))[which.max(table(temp2$chrom))])]] = list(seq = temp2, tpt = tpt2)
}
# ## Getting rf matrices
# rfs.inter = lapply(LGS.inter, function(x) rf_list_to_matrix(x$tpt))
# ## Plotting rf matrices
# #png(file="UPGMA_Genomic.png", width = 1600, height = 800)
# par(mfrow=c(2,4))
# for (i in 1:length(rfs.inter)){plot(rfs.inter[[i]])}
# mtext("Linkage groups with\nUPGMA + genomic information\n(discarding scaffolds)", cex = 2, side = 3, line = -63, at = 0.88, outer = TRUE)
# par(mfrow=c(1,1))
# #dev.off()

## Making groups only with genomic information (tends to bring more markers, but also discards scaffolds)
LGS.genomic = vector("list", 7)
for (i in 1:7){
  tempseq1 = make_seq_mappoly(dat, arg = paste0("seq",i), genomic.info = 1)
  mrks = intersect(tempseq1$seq.mrk.names, seq.init$seq.mrk.names)
  tempseq = make_seq_mappoly(dat, arg = mrks)
  temptpt = make_pairs_mappoly(all.rf.pairwise, input.seq = tempseq)
  rffilt = rf_snp_filter(input.twopt = temptpt, diagnostic.plot = FALSE)
  temptpt2 = make_pairs_mappoly(temptpt, input.seq = rffilt)
  LGS.genomic[[as.numeric(unique(rffilt$chrom))]] = list(seq = rffilt, tpt = temptpt2)
}
# ## Getting rf matrices
# rfs.genomic = lapply(LGS.genomic, function(x) rf_list_to_matrix(x$tpt))
# ## Plotting rf matrices
# #png(file="Genomic.png", width = 1600, height = 800)
# par(mfrow=c(2,4))
# for (i in 1:length(rfs.genomic)){plot(rfs.genomic[[i]])}
# mtext("Linkage groups with\ngenomic information\n(discarding scaffolds)", cex = 2, side = 3, line = -63, at = 0.88, outer = TRUE)
# par(mfrow=c(1,1))
# #dev.off()

## Making groups with UPGMA (rf) information (includes scaffolds)
LGS.upgma=vector("list", 7)
for(j in 1:7){
  temp1 = make_seq_mappoly(grs, j)
  tpt = make_pairs_mappoly(all.rf.pairwise, input.seq = temp1)
  temp2 = rf_snp_filter(input.twopt = tpt, diagnostic.plot = FALSE)
  tpt2 = make_pairs_mappoly(tpt, input.seq = temp2)
  LGS.upgma[[as.numeric(names(table(temp2$chrom))[which.max(table(temp2$chrom))])]] = list(seq = temp2, tpt = tpt2)
}
# ## Getting rf matrices
# rfs = lapply(LGS.upgma, function(x) rf_list_to_matrix(x$tpt))
# ## Plotting rf matrices
# #png(file="UPGMA.png", width = 1600, height = 800)
# par(mfrow=c(2,4))
# for (i in 1:length(rfs)){plot(rfs[[i]])}
# mtext("Linkage groups with\nUPGMA information\n(including scaffolds)", cex = 2, side = 3, line = -63, at = 0.88, outer = TRUE)
# par(mfrow=c(1,1))
# #dev.off()

## Comparing number of markers in each group
comp = data.frame(UPGMA_Genomic = unlist(lapply(LGS.inter, function(x) length(x$seq$seq.num))),
                  Genomic = unlist(lapply(LGS.genomic, function(x) length(x$seq$seq.num))),
                  UPGMA = unlist(lapply(LGS.upgma, function(x) length(x$seq$seq.num))))
comp


```

We see that all three are very similar and for the purpose of this example lets just take the upgma grouping and run with that.

```{r}
LGS<-LGS.upgma
```

# HMM chain

The HMM chain is used to order the markers. Below is a example of just ordering one chromosome. Later we will show how to parallelize this task by chromosome. (NOTE the map is only running quickly here because of the small number of markers and the good quality of the data)

```{r}
single_chrom <- est_rf_hmm_sequential(input.seq = LGS[[1]]$seq,
                               start.set = 3,
                               thres.twopt = 10,
                               thres.hmm = 50,
                               extend.tail = 30,
                               twopt = LGS[[1]]$tpt,
                               verbose = TRUE,
                               phase.number.limit = 20,
                               sub.map.size.diff.limit = 5)
```

```{r}
plot(single_chrom)
```

## Parallel construction of all 7 LGs

```{r}
## Performing parallel computation of hidden markov 
phasing_and_hmm_rf <- function(X){
  fl <- paste0("output_map_ch_", unique(X$seq$chrom), ".txt")
  sink(fl)
  map <- est_rf_hmm_sequential(input.seq = X$seq,
                               start.set = 3,
                               thres.twopt = 10,
                               thres.hmm = 50,
                               extend.tail = 30,
                               twopt = X$tpt,
                               verbose = TRUE,
                               phase.number.limit = 20,
                               sub.map.size.diff.limit = 5) 
  sink()
  return(map)
}


# 5% genotyping error
my.error.func<-function(X){
  x<-est_full_hmm_with_global_error(input.map = X, 
                                    error = 0.05, 
                                    tol = 10e-4, 
                                    verbose = FALSE)
  return(x)
}
#upgma
#hidden markov based off physical location
cl <- parallel::makeCluster(7)
parallel::clusterEvalQ(cl, require(mappoly))
parallel::clusterExport(cl, "dat")
MAPs <- parallel::parLapply(cl,LGS, phasing_and_hmm_rf)
parallel::stopCluster(cl)

plot_map_list(MAPs, col = "ggstyle")
# account for error
cl <- parallel::makeCluster(7)
parallel::clusterEvalQ(cl, require(mappoly))
parallel::clusterExport(cl, "dat")
MAPs.err <- parallel::parLapply(cl,MAPs,my.error.func)
parallel::stopCluster(cl)

par(mfrow=c(2,1))
plot_map_list(MAPs, col = "ggstyle")
plot_map_list(MAPs.err, col = "ggstyle")
```

After constructing the maps a 5% global error rate is applied to the chain to allow for the chain to make small changes that correct for possible genotyping error. The difference is seen the the picture above where the second map is shorter indicating a better map.

```{r}
plot_genome_vs_map(MAPs.err, same.ch.lg = T)
summary_maps(MAPs.err)
```

# Genotypic probabilites for use in QTL mapping

Genotypic probabilites will be used in QTL mapping thus we need to save this for use in next session

```{r}
genoprob.err <- vector("list", 7)
for(i in 1:7)
   genoprob.err[[i]] <- calc_genoprob_error(input.map = MAPs.err[[i]], error = 0.05)

saveRDS(genoprob.err, file="genoprob.err.RDS")
```

# working in polymapR

polymapR has the advantage of lower computational overhead, however less automatic and clustering may be more difficult. However if you want to work with this platform, you can use export the mappoly.data

```{r}
library(polymapR)

ALL_dosages =  export_data_to_polymapR(dat)

screened_data <- screen_for_duplicate_individuals(dosage_matrix = ALL_dosages, 
                                                   cutoff = 0.95, 
                                                   plot_cor = T)

PCA_progeny(screened_data)
```

```{r}
data=read.csv("data/BExMG_subset.csv")

removed_dups=cbind(data[,1:5],screened_data[,3:ncol(screened_data)])

```

```{r}
dat = read_geno_csv(file="data/BExMG_subset_removed_dup.csv", ploidy = 4)
## Filtering dataset by marker
dat = filter_missing(input.data = dat, type = "marker", 
                      filter.thres = 0.05, inter = F)
## Filtering dataset by individual
dat = filter_missing(input.data = dat, type = "individual", 
                      filter.thres = 0.05, inter = F)
## Filter according to segregation test
pval.bonf = 0.05/dat$n.mrk
mrks.chi.filt = filter_segregation(dat, chisq.pval.thres =  pval.bonf, inter = F)
seq.init = make_seq_mappoly(mrks.chi.filt)
## Accounting for recombination fractions
all.rf.pairwise = est_pairwise_rf(input.seq = seq.init, ncpus = 6)
## Getting group information (UPGMA)
grs = group_mappoly(input.mat = mat,
                     expected.groups = 7,
                     comp.mat = TRUE, 
                     inter = F)
grs

## Making groups with intersection between UPGMA (rf) + genomic information (discards scaffolds)

LGS.inter=vector("list", 7)
for(j in 1:7){
  temp1 = make_seq_mappoly(grs, j, genomic.info=1)
  tpt = make_pairs_mappoly(all.rf.pairwise, input.seq = temp1)
  temp2 = rf_snp_filter(input.twopt = tpt, diagnostic.plot = FALSE)
  tpt2 = make_pairs_mappoly(tpt, input.seq = temp2)
  LGS.inter[[as.numeric(names(table(temp2$chrom))[which.max(table(temp2$chrom))])]] = list(seq = temp2, tpt = tpt2)
}


## Making groups only with genomic information (tends to bring more markers, but also discards scaffolds)
LGS.genomic = vector("list", 7)
for (i in 1:7){
  tempseq1 = make_seq_mappoly(dat, arg = paste0("seq",i), genomic.info = 1)
  mrks = intersect(tempseq1$seq.mrk.names, seq.init$seq.mrk.names)
  tempseq = make_seq_mappoly(dat, arg = mrks)
  temptpt = make_pairs_mappoly(all.rf.pairwise, input.seq = tempseq)
  rffilt = rf_snp_filter(input.twopt = temptpt, diagnostic.plot = FALSE)
  temptpt2 = make_pairs_mappoly(temptpt, input.seq = rffilt)
  LGS.genomic[[as.numeric(unique(rffilt$chrom))]] = list(seq = rffilt, tpt = temptpt2)
}


## Making groups with UPGMA (rf) information (includes scaffolds)
LGS.upgma=vector("list", 7)
for(j in 1:7){
  temp1 = make_seq_mappoly(grs, j)
  tpt = make_pairs_mappoly(all.rf.pairwise, input.seq = temp1)
  temp2 = rf_snp_filter(input.twopt = tpt, diagnostic.plot = FALSE)
  tpt2 = make_pairs_mappoly(tpt, input.seq = temp2)
  LGS.upgma[[as.numeric(names(table(temp2$chrom))[which.max(table(temp2$chrom))])]] = list(seq = temp2, tpt = tpt2)
}


## Comparing number of markers in each group
comp = data.frame(UPGMA_Genomic = unlist(lapply(LGS.inter, function(x) length(x$seq$seq.num))),
                  Genomic = unlist(lapply(LGS.genomic, function(x) length(x$seq$seq.num))),
                  UPGMA = unlist(lapply(LGS.upgma, function(x) length(x$seq$seq.num))))
comp

LGS<-LGS.upgma

## Performing parallel computation of hidden markov 
phasing_and_hmm_rf <- function(X){
  fl <- paste0("output_map_ch_", unique(X$seq$chrom), ".txt")
  sink(fl)
  map <- est_rf_hmm_sequential(input.seq = X$seq,
                               start.set = 3,
                               thres.twopt = 10,
                               thres.hmm = 50,
                               extend.tail = 30,
                               twopt = X$tpt,
                               verbose = TRUE,
                               phase.number.limit = 20,
                               sub.map.size.diff.limit = 5) 
  sink()
  return(map)
}


# 5% genotyping error
my.error.func<-function(X){
  x<-est_full_hmm_with_global_error(input.map = X, 
                                    error = 0.05, 
                                    tol = 10e-4, 
                                    verbose = FALSE)
  return(x)
}
#upgma
#hidden markov based off physical location
cl <- parallel::makeCluster(7)
parallel::clusterEvalQ(cl, require(mappoly))
parallel::clusterExport(cl, "dat")
MAPs_nodup <- parallel::parLapply(cl,LGS, phasing_and_hmm_rf)
parallel::stopCluster(cl)

# account for error
cl <- parallel::makeCluster(7)
parallel::clusterEvalQ(cl, require(mappoly))

parallel::clusterExport(cl, "dat")
MAPs_nodup.err <- parallel::parLapply(cl,MAPs_nodup,my.error.func)
parallel::stopCluster(cl)

par(mfrow=c(2,1))
plot_map_list(MAPs.err, col = "ggstyle")
plot_map_list(MAPs_nodup.err, col = "ggstyle")

summary_maps(MAPs.err)
summary_maps(MAPs_nodup.err)
```
