---
title: "QTL Example"
author: "Jeekin Lau"
date: "Updated: 11/30/2022"
output:   
  html_document:
    toc: true
    toc_float: 
      collapsed: false
---


# QTLpoly for QTL analysis

MAPpoly's sister software QTLpoly was made to take the outputs of MAPpoly to do QTL mapping. There are two main ways in QTLpoly REMIM random-effect multiple interval mapping and FEIM fixed effect interval mapping.

## Import data

First we need the to read in the genotypic probabilites file `genoprob.err` object that was generated in the previous linkage mapping session.

```{r}
library(qtlpoly)

genoprob.err = readRDS(file="genoprob.err.RDS")

pheno = read.csv("../data/blackspot_cercospora_defoliation_RRD.csv", row.names = 1)

data = read_data(ploidy = 4, geno.prob = genoprob.err, pheno = pheno, step = 1)


# this step takes a long time so do not run import it
#data.sim = simulate_qtl(data = data, mu = 0, h2.qtl = NULL, var.error = 1, n.sim = 1000, missing = TRUE, seed = 123) 
#score.null = null_model(data = data.sim$results, n.clusters = 7, plot = NULL)
```

## Didactic example of one trait for teaching REMIM

The REMIM method consists of a few steps: this is a simplification for more math terms go here: <https://guilherme-pereira.github.io/QTLpoly/1-tutorial#Perform_QTL_detection> 1. Null model: model starts with no QTL 2. Forward search: QTL are added one at a time conditional to any previous if any QTL already in the model 3. Model optimization: each QTL is then tested again conditional to all other QTL under a more stringent threshold - steps 2 and 3 are repeated until no more QTL are added and dropped from model. 4. QTL profiling: score statistics are updated conditional to QTL that are retained

### 1. Null model

```{r}
null.mod <- null_model(data = data, pheno.col = 1 ,n.clusters = 4, plot = "null")
print(null.mod)
```

### 2. Forward search

```{r}
search.mod <- search_qtl(data = data, model = null.mod, w.size = 15, sig.fwd = 0.01, n.clusters = 4, plot = "search")
print(search.mod)
```

### 3. Optimize QTL

```{r}
optimize.mod <- optimize_qtl(data = data, model = search.mod, sig.bwd = 1e-04,  n.clusters = 4, plot = "optimize")
print(optimize.mod)
```

### 3. Profile QTL

```{r}
profile.mod <- profile_qtl(data = data, model = optimize.mod, d.sint = 1.5,  polygenes = FALSE, n.clusters = 4, plot = "profile")
```

```{r}
print(profile.mod)
print(profile.mod, sint="lower")
print(profile.mod, sint="upper")
```

### Plotting the QTL profiles

```{r}
plot_profile(data=data,model=profile.mod, sup.int = T) 
```

### Run all of the traits

Because the score.null object (genome wide significance levels) take so long to calculate, I have provided done it in advance. If you have a new map, or if you have a new experiment, you will need to recalculate.

```{r}
score.null = readRDS("../data/score.null.RDS")

remim.mod = remim(data, score.null = score.null, w.size = 15, sig.fwd = 0.2, sig.bwd = 0.05, n.clusters = 7)
plot_profile(data, remim.mod, sup.int = T)
plot_profile(data, remim.mod, sup.int = T, grid=T)
plot_sint(data,remim.mod)
```

`fit_model()` calculates all the QTL's variance components and estimates heritability of the QTL (comparable to the R-squared of the QTL)

```{r}
fitted.mod <- fit_model(data = data, model = remim.mod, probs = "joint", polygenes = "none")
summary(fitted.mod) #some reason this does not display right please run in the console to see results
```

### qtl_effects

`qtl_effects()` shows the user for each QTL what parental homolog is responsible for increasing or decreasing the phenotypic average.

```{r}
est.effects <- qtl_effects(ploidy = 4, fitted = fitted.mod)
plot(est.effects)
```

## Save results for visualization in Viewpoly

What we need to do is now save all the objects so that we can now look at a more beautiful output for mining your results.

```{r}
# linkage map file
save(MAPs.err,file = "../viewpoly_files/MAPs.err.RData")
save(data, file = "../viewpoly_files/QTLpoly_data.RData")
save(remim.mod, file = "../viewpoly_files/QTLpoly_remim.mod.RData")
save(est.effects, file = "../viewpoly_files/QTLpoly_est.effects.RData")
save(fitted.mod, file = "../viewpoly_files/QTLpoly_fitted.mod.RData") 
```





